Unable to launch virt-handler v0.20.4 on CoreOS with selinux

/kind bug


**What happened**:
I started kubevirt v0.20.4 using some instructions from https://kubevirt.io/quickstart_minikube/ on a Kubernetes v1.15 cluster running on bare-metal nodes with CoreOS 1883.0.0 (~1yo ...).

**What you expected to happen**:
`virt-handler` fails to start with:
```
{"component":"virt-handler","hostname":"carmd-er-lxtb0cc47a4c27b6","level":"info","pos":"virt-handler.go:187","timestamp":"2019-09-04T06:02:51.362115Z"}
panic: failed to list labels:  

goroutine 1 [running]:
main.(*virtHandlerApp).Run(0xc0001f6000)
	cmd/virt-handler/virt-handler.go:205 +0x1ad6
main.main()
	cmd/virt-handler/virt-handler.go:457 +0x6e
```

**How to reproduce it (as minimally and precisely as possible)**:
I think running kubevirt on CoreOS should be enough, but hard to say.

**Anything else we need to know?**:
It looks like the `selinux/labels.go` is calling `semanage`, and I couldn't find it in the image.

**Environment**:
- KubeVirt version (use `virtctl version`): v0.20.4
- Kubernetes version (use `kubectl version`): v1.15.1
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: bare-metal x86-64 servers
- OS (e.g. from /etc/os-release): CoreOS 1883.0.0 (w/ ACS patch)
- Kernel (e.g. `uname -a`): 4.18.5-coreos
- Install tools:
- Others:

@slintes , sounds good, didn't knew that was expected.

@cynepco3hahue , I tried the PR #2665 and I see the following:
```
{"component":"virt-handler","hostname":"carmd-er-lxtb0cc47a4c2598","level":"info","pos":"virt-handler.go:187","timestamp":"2019-09-04T13:15:55.797845Z"}
panic: failed to list labels: out: "", error: could not find binary semanage
goroutine 1 [running]:
main.(*virtHandlerApp).Run(0xc000316100)
	cmd/virt-handler/virt-handler.go:205 +0x1ba0
main.main()
	cmd/virt-handler/virt-handler.go:461 +0x6e
```
Seems like we would need to ship `semanage` with the image, provided by package `policycoreutils-python-utils`.

I was able to do that in https://github.com/CoRfr/kubevirt/commit/a1c91af9b0e658c0730cca4d86dabb216ae9d372 , but I still get `error: could not find binary semanage`.

I suspect there is some error in the lookupPath function:
```
Looking up semanage
Trying /proc/1/root/sbin/semanage
Error stat /proc/1/root/sbin/semanage: no such file or directory
Trying /proc/1/root/usr/sbin/semanage
Error stat /proc/1/root/usr/sbin/semanage: no such file or directory
Trying /proc/1/root/bin/semanage
Error stat /proc/1/root/bin/semanage: no such file or directory
Trying /proc/1/root/usr/bin/semanage
Error stat /proc/1/root/usr/bin/semanage: no such file or directory
panic: failed to list labels: out: "", error: could not find binary semanage
```

--- a/pkg/virt-handler/selinux/BUILD.bazel
+++ b/pkg/virt-handler/selinux/BUILD.bazel
@@ -5,6 +5,7 @@ go_library(
     srcs = ["labels.go"],
     importpath = "kubevirt.io/kubevirt/pkg/virt-handler/selinux",
     visibility = ["//visibility:public"],
+    deps = ["//staging/src/kubevirt.io/client-go/log:go_default_library"],
 )
 
 go_test(

--- a/pkg/virt-handler/selinux/labels.go
+++ b/pkg/virt-handler/selinux/labels.go
@@ -7,6 +7,8 @@ import (
 	"os/exec"
 	"path/filepath"
 	"strings"
+
+	"kubevirt.io/client-go/log"
 )
 
 const procOnePrefix = "/proc/1/root"
@@ -54,6 +56,17 @@ func (se *SELinuxImpl) IsPresent() (present bool, err error) {
 		return false, nil
 	}
 	return true, nil
+}
+
+func (se *SELinuxImpl) IsPermissive() (present bool, err error) {
+	out, err := se.execute("getenforce", se.Paths)
+	if err != nil {
+		return false, err
+	}
+	if strings.Contains(string(out), "Permissive") {
+		return true, nil
+	}
+	return false, nil
 }
 
 func lookupPath(binary string, prefix string, paths []string) (string, bool, error) {
@@ -107,6 +120,10 @@ func copyPolicy(policyName string, dir string) (err error) {
 func (se *SELinuxImpl) Label(label string, dir string) error {
 	dir = strings.TrimRight(dir, "/") + "(/.*)?"
 	if out, err := se.execute("semanage", se.Paths, "fcontext", "-a", "-t", label, dir); err != nil {
+		if perm, _ := se.IsPermissive(); perm {
+			log.Log.Warningf("Permissive mode, ignoring 'semanage' failure: out: %q, error: %v", string(out), err)
+			return nil
+		}
 		return fmt.Errorf("failed to set label for directory %v: out: %q, error: %v", dir, string(out), err)
 	}
 	return nil
@@ -117,6 +134,10 @@ func (se *SELinuxImpl) IsLabeled(dir string) (bool, error) {
 	dir = strings.TrimRight(dir, "/") + "(/.*)?"
 	out, err := se.execute("semanage", se.Paths, "fcontext", "-l")
 	if err != nil {
+		if perm, _ := se.IsPermissive(); perm {
+			log.Log.Warningf("Permissive mode, ignoring 'semanage' failure: out: %q, error: %v", string(out), err)
+			return false, nil
+		}
 		return false, fmt.Errorf("failed to list labels: out: %q, error: %v", string(out), err)
 	}
 	if strings.Contains(string(out), dir) {
@@ -129,20 +150,28 @@ func (se *SELinuxImpl) IsLabeled(dir string) (bool, error) {
 func (se *SELinuxImpl) Restore(dir string) error {
 	dir = strings.TrimRight(dir, "/") + "/"
 	if out, err := se.execute("restorecon", se.Paths, "-r", "-v", dir); err != nil {
+		if perm, _ := se.IsPermissive(); perm {
+			log.Log.Warningf("Permissive mode, ignoring 'restorecon' failure: out: %q, error: %v", string(out), err)
+			return nil
+		}
 		return fmt.Errorf("failed to set selinux permissions: out: %q, error: %v", string(out), err)
 	}
 	return nil
 }
 
-func (*SELinuxImpl) InstallPolicy(dir string) (err error) {
+func (se *SELinuxImpl) InstallPolicy(dir string) (err error) {
 	for _, policyName := range POLICY_FILES {
 		fileDest := dir + "/" + policyName + ".cil"
 		err := copyPolicy(policyName, dir)
 		if err != nil {
 			return fmt.Errorf("failed to copy policy %v - err: % v", fileDest, err)
 		}
-		_, err = exec.Command("/usr/bin/chroot", "--mount", "/proc/1/ns/mnt", "exec", "--", "/usr/sbin/semodule", "-i", fileDest).CombinedOutput()
+		out, err := exec.Command("/usr/bin/chroot", "--mount", "/proc/1/ns/mnt", "exec", "--", "/usr/sbin/semodule", "-i", fileDest).CombinedOutput()
 		if err != nil {
+			if perm, _ := se.IsPermissive(); perm {
+				log.Log.Warningf("Permissive mode, ignoring 'semodule' failure: out: %q, error: %v", string(out), err)
+				return nil
+			}
 			return fmt.Errorf("failed to install policy %v - err: % v", fileDest, err)
 		}
 	}

--- a/pkg/virt-handler/selinux/labels_test.go
+++ b/pkg/virt-handler/selinux/labels_test.go
@@ -47,6 +47,15 @@ var _ = Describe("selinux", func() {
 			Expect(err).To(BeNil())
 			Expect(present).To(BeFalse())
 		})
+		It("should detect that it is enabled if getenforce returns Permissive", func() {
+			touch(filepath.Join(tempDir, "/usr/bin", "getenforce"))
+			selinux.execFunc = func(binary string, args ...string) (bytes []byte, e error) {
+				return []byte("Permissive"), nil
+			}
+			present, err := selinux.IsPresent()
+			Expect(err).To(BeNil())
+			Expect(present).To(BeTrue())
+		})
 		It("should detect that it is enabled if getenforce does not return Disabled", func() {
 			touch(filepath.Join(tempDir, "/usr/bin", "getenforce"))
 			selinux.execFunc = func(binary string, args ...string) (bytes []byte, e error) {
@@ -91,6 +100,18 @@ var _ = Describe("selinux", func() {
 		It("should fail if semanage exists but the command fails", func() {
 			touch(filepath.Join(tempDir, "/usr/bin", "semanage"))
 			selinux.execFunc = func(binary string, args ...string) (bytes []byte, e error) {
+				return nil, fmt.Errorf("I failed")
+			}
+			_, err := selinux.IsLabeled("whatever")
+			Expect(err).To(HaveOccurred())
+		})
+
+		It("should succeed if semanage exists and the command fails, but we are in Permissive mode", func() {
+			touch(filepath.Join(tempDir, "/usr/bin", "semanage"))
+			selinux.execFunc = func(binary string, args ...string) (bytes []byte, e error) {
+				if binary == "/usr/sbin/getenforce" {
+					return []byte("Permissive"), nil
+				}
 				return nil, fmt.Errorf("I failed")
 			}
 			_, err := selinux.IsLabeled("whatever")

